helix-lsp/src/lib.rs

```diff
@@ -55,55 +55,91 @@ pub mod util {
     use super::*;
     use helix_core::{Range, Rope, Transaction};
 
+    /// Converts [`lsp::Position`] to a position in the document.
+    ///
+    /// Returns `None` if position exceeds document length or an operation overflows.
     pub fn lsp_pos_to_pos(
         doc: &Rope,
         pos: lsp::Position,
         offset_encoding: OffsetEncoding,
-    ) -> usize {
+    ) -> Option<usize> {
+        let max_line = doc.lines().count().saturating_sub(1);
+        let pos_line = pos.line as usize;
+        let pos_line = if pos_line > max_line {
+            return None;
+        } else {
+            pos_line
+        };
         match offset_encoding {
             OffsetEncoding::Utf8 => {
-                let line = doc.line_to_char(pos.line as usize);
-                line + pos.character as usize
+                let max_char = doc
+                    .line_to_char(max_line)
+                    .checked_add(doc.line(max_line).len_chars())?;
+                let line = doc.line_to_char(pos_line);
+                let pos = line.checked_add(pos.character as usize)?;
+                if pos <= max_char {
+                    Some(pos)
+                } else {
+                    None
+                }
             }
             OffsetEncoding::Utf16 => {
-                let line = doc.line_to_char(pos.line as usize);
+                let max_char = doc
+                    .line_to_char(max_line)
+                    .checked_add(doc.line(max_line).len_chars())?;
+                let max_cu = doc.char_to_utf16_cu(max_char);
+                let line = doc.line_to_char(pos_line);
                 let line_start = doc.char_to_utf16_cu(line);
-                doc.utf16_cu_to_char(line_start + pos.character as usize)
+                let pos = line_start.checked_add(pos.character as usize)?;
+                if pos <= max_cu {
+                    Some(doc.utf16_cu_to_char(pos))
+                } else {
+                    None
+                }
             }
         }
     }
+
+    /// Converts position in the document to [`lsp::Position`].
+    ///
+    /// Returns `None` if position exceeds the document length or an operation overflows.
     pub fn pos_to_lsp_pos(
         doc: &Rope,
         pos: usize,
         offset_encoding: OffsetEncoding,
-    ) -> lsp::Position {
+    ) -> Option<lsp::Position> {
```

Comment:
I think we can avoid using an Option<> for `pos_to_lsp_pos` because the pos is coming from trusted input (selections/transactions always have to be valid)

Reply:
This should significantly simplify how much `Option` handling we need to do since we mostly use `pos_to_lsp_pos`

Reply:
Done, reverted `pos_to_lsp_pos` to return just `usize`

Reply:
I'm wondering if the function should saturate the substitutions? We still might need to handle this overflow somehow.

Reply:
I think it's fine to just let it panic. We may want to set up some general `debug_assert!`s  that run on each event to verify selections though, that would be a good test for correctness

Reply:
Ok, I got rid of the saturating ops and bound check in `pos_to_lsp_pos`.